File: solutions/question_63_unique_path_2.rb

# Follow up for "Unique Paths":
#
# Now consider if some obstacles are added to the grids. How many unique paths would there be?
#
# An obstacle and empty space is marked as 1 and 0 respectively in the grid.
#
# For example,
# There is one obstacle in the middle of a 3x3 grid as illustrated below.
#
# [
#   [0,0,0],
#   [0,1,0],
#   [0,0,0]
# ]
# The total number of unique paths is 2.
#
# Note: m and n will be at most 100.

# @param {Integer[][]} obstacle_grid
# @return {Integer}
def unique_paths_with_obstacles(obstacle_grid)
  r = obstacle_grid
  m = r.size
  n = r[0].size || 0
  return 0 if m.zero? || r[m - 1][n - 1] == 1

  ri = r[m - 1].rindex(1)
  if ri.nil?
    r[m - 1] = Array.new(n, 1)
  else
    (0..n - 1).each { |j| r[m - 1][j] = j <= ri ? 0 : 1 }
  end

  (0..m - 2).reverse_each do |i|
    if r[i][n - 1] == 1
      (0..i).each { |j| r[j][n - 1] = 0 }
      break
    end
    r[i][n - 1] = 1
  end

  (0..m - 2).reverse_each do |i|
    (0..n - 2).reverse_each do |j|
      r[i][j] = if r[i][j] == 1
                  0
                else
                  r[i][j + 1] + r[i + 1][j]
                end
    end
  end
  r[0][0]
end

og = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
p unique_paths_with_obstacles(og)

File: _site/solutions/question_63_unique_path_2.py
# Follow up for "Unique Paths":
#
# Now consider if some obstacles are added to the grids. How many unique paths would there be?
#
# An obstacle and empty space is marked as 1 and 0 respectively in the grid.
#
# For example,
# There is one obstacle in the middle of a 3x3 grid as illustrated below.
#
# [
#   [0,0,0],
#   [0,1,0],
#   [0,0,0]
# ]
# The total number of unique paths is 2.
#
# Note: m and n will be at most 100.

# @param {Integer[][]} obstacle_grid
# @return {Integer}
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        ob = obstacleGrid
        m, n = len(ob), len(ob[0])
        if not m or ob[m-1][n-1] == 1: return 0

        for i in reversed(range(n)):
            if ob[m-1][i] == 0:
                ob[m-1][i] = 1
            else:
                for j in range(i + 1):
                    ob[m-1][j] = 0
                break

        for i in reversed(range(0, m - 1)): # xrange from 1
            if ob[i][n - 1] == 0:
                ob[i][n - 1] = 1
            else:
                for j in range(i + 1):
                    ob[j][n - 1] = 0
                break

        for i in reversed(range(m - 1)):
            for j in reversed(range(n - 1)):
                ob[i][j] = 0 if ob[i][j] == 1 else ob[i+1][j] + ob[i][j+1]

        return ob[0][0]

ob = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
s = Solution()
print (s.uniquePathsWithObstacles(ob))

